Source/DataSources/I3SDataSource.js

```diff
@@ -0,0 +1,4601 @@
+.##...####.######...##.....##.##.###.##.######......##....########.....##...
+.##....##..##.......##.....##.##.....##.##..........##....##...##......##...
+.##....##..##.......##.....##.##.....##.##..........##....##....##.....##...
+..######...########..#######..##.....##.########....##....##.....##....##...
+*/
+
+/**
+ * This class implements an I3S Geometry, in Cesium, each I3SGeometry
+ * generates an in memory b3dm so be used as content for a Cesium3DTile
+ * @private
+ * @alias I3SGeometry
+ * @constructor
+ * @param {I3SNode} [parent] The parent of that geometry
+ * @param {String} [uri] The uri to load the data from
+ */
+function I3SGeometry(parent, uri) {
+  this._parent = parent;
+  this._dataSource = parent._dataSource;
+  this._layer = parent._layer;
+  this._uri = uri;
+  var query = "";
+  if (this._dataSource._query && this._dataSource._query !== "") {
+    query = "?" + this._dataSource._query;
+  }
+
+  if (this._parent._nodeIndex) {
+    this._completeUriWithoutQuery =
+      this._layer._completeUriWithoutQuery +
+      "/nodes/" +
+      this._parent._data.mesh.geometry.resource +
+      "/" +
+      this._uri;
+    this._completeUri = this._completeUriWithoutQuery + query;
+  } else {
+    this._completeUriWithoutQuery =
+      this._parent._completeUriWithoutQuery + "/" + this._uri;
+    this._completeUri = this._completeUriWithoutQuery + query;
+  }
+}
+
+Object.defineProperties(I3SGeometry.prototype, {
+  /**
+   * Gets the uri for the geometry.
+   * @memberof I3SGeometry.prototype
+   * @type {string}
+   */
+  uri: {
+    get: function () {
+      return this._uri;
+    },
+  },
+  /**
+   * Gets the complete uri for the geometry.
+   * @memberof I3SGeometry.prototype
+   * @type {string}
+   */
+  completeUri: {
+    get: function () {
+      return this._completeUri;
+    },
+  },
+  /**
+   * Gets the I3S data for this object.
+   * @memberof I3SGeometry.prototype
+   * @type {Object}
+   */
+  data: {
+    get: function () {
+      return this._data;
+    },
+  },
+});
+
+/**
+ * Loads the content.
+ * @returns {Promise<void>} a promise that is resolved when the geometry data is loaded
+ */
+I3SGeometry.prototype.load = function () {
+  var that = this;
+  return this._dataSource._loadBinary(
+    this._completeUri,
+    function (data, resolve) {
+      that._data = data;
+      resolve();
+    },
+    function (reject) {
+      reject();
+    }
+  );
+};
+
+function dot(a, b) {
+  return a.x * b.x + a.y * b.y + a.z * b.z;
+}
+function cross(a, b) {
+  return new Cartesian3(
+    a.y * b.z - a.z * b.y,
+    a.z * b.x - a.x * b.z,
+    a.x * b.y - a.y * b.x
+  );
+}
+function subtract(a, b) {
+  return new Cartesian3(a.x - b.x, a.y - b.y, a.z - b.z);
+}
+function scale(v, scalar) {
+  return new Cartesian3(v.x * scalar, v.y * scalar, v.z * scalar);
+}
+function normalize(a) {
+  return Cartesian3.magnitude(a) > 0
+    ? scale(a, 1 / Cartesian3.magnitude(a))
+    : a;
+}
+
+function sameSide(p1, p2, a, b) {
```

Comment:
Any new functionality for a class should go into the class's file or in [`Math.js`](https://github.com/CesiumGS/cesium/blob/main/Source/Core/Math.js), if appropriate.